File: fictional-shop-checkout/example-java/src/main/java/org/example/domain/value/DirectItem.java
package org.example.domain.value;

import org.example.domain.DomainException;

/**
 * 商品情報（直打ち）
 */
public class DirectItem {

  /**
   * 商品の金額
   */
  private final int price;

  /**
   * 商品の種別番号
   */
  private final String typeId;

  private static final int PRICE_MAX_VALUE = 1000000;

  public DirectItem(String typeId, int price) {
    this.typeId = typeId;
    this.price = price;
  }
  public DirectItem(String typeIdAndPrice) {
    var values = typeIdAndPrice.split(":");
    var rawTypeId = values[0];
    var rawPrice = values[1];

    this.typeId = rawTypeId;
    this.price = Integer.parseInt(rawPrice);

    if (this.price > PRICE_MAX_VALUE) {
      throw new DomainException("直打ち商品：金額が1,000,000円を超えているため、扱うことができません。");
    }

  }

  public int getPrice() {
    return price;
  }
}

File: fictional-shop-checkout/example-kotlin/src/main/kotlin/org/example/domain/value/DirectItem.kt
package org.example.domain.value

import org.example.domain.DomainException

/**
 * 商品情報（直打ち）
 */
class DirectItem {
    /** 商品の金額 */
    val price: Int
    /** 種別番号 */
    val typeId: String

    companion object {
        private const val PRICE_MAX_VALUE = 1_000_000
    }

    constructor(typeId: String, price: Int) {
        this.typeId = typeId
        this.price = price
        if (this.price > PRICE_MAX_VALUE) {
            throw DomainException("直打ち商品：金額が1,000,000円を超えているため、扱うことができません。")
        }
    }

    constructor(typeIdAndPrice: String) {
        val values = typeIdAndPrice.split(":")
        val rawTypeId = values[0]
        val rawPrice = values[1]
        this.typeId = rawTypeId
        this.price = rawPrice.toInt()
        if (this.price > PRICE_MAX_VALUE) {
            throw DomainException("直打ち商品：金額が1,000,000円を超えているため、扱うことができません。")
        }
    }
}